src/Symfony/Component/Uri/Uri.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Uri;
+
+use Symfony\Component\Uri\Exception\InvalidUriException;
+
+/**
+ * Parses a URI and allows to resolve relative URIs, as defined
+ * in RFC 3986 (https://tools.ietf.org/html/rfc3986).
+ *
+ * @experimental
+ *
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+final class Uri implements \Stringable
+{
+    private const URI_GLOBAL_REGEX = '/^(?:(?P<scheme>[^:\/?#]+):)?(?:\/\/(?P<authority>[^\/?#]*))?(?P<path>[^?#]*)(?:\?(?P<query>[^#]*))?(?:#(?P<fragment>.*))?$/';
+    private const URI_AUTHORITY_REGEX = '/^(?:(?P<user>[^:@]*)(?::(?P<pass>[^@]*))?@)?(?P<host>[^:]*)(?::(?P<port>\d*))?$/';
+
+    public function __construct(
+        public string $scheme,
+        #[\SensitiveParameter]
+        public ?string $user = null,
+        #[\SensitiveParameter]
+        public ?string $password = null,
+        public ?string $host = null,
+        public ?int $port = null,
+        public ?string $path = null,
+        public ?QueryString $query = null,
+        public ?string $fragment = null,
+        public ?FragmentTextDirective $fragmentTextDirective = null,
+    ) {
+    }
+
+    /**
+     * Parses a URL.
+     *
+     * The `user` and `pass` keys are url-decoded automatically when parsing.
+     *
+     * @throws InvalidUriException
+     */
+    public static function parse(string $uri): static
+    {
+        preg_match(self::URI_GLOBAL_REGEX, $uri, $matches);
+        if (!$matches || !isset($matches['scheme']) || '' === $matches['scheme']) {
+            throw new InvalidUriException($uri);
+        }
+
+        if (isset($matches['authority'])) {
+            preg_match(self::URI_AUTHORITY_REGEX, $matches['authority'], $authMatches);
+
+            $matches = array_merge($matches, $authMatches);
+            unset($matches['authority']);
+        }
+
+        $matches = array_filter($matches, static fn (string $value): bool => '' !== $value);
+
+        return new static(
+            $matches['scheme'],
+            isset($matches['user']) ? rawurldecode($matches['user']) : null,
+            isset($matches['pass']) ? rawurldecode($matches['pass']) : null,
+            $matches['host'] ?? null,
+            $matches['port'] ?? null,
+            $matches['path'] ?? null,
+            isset($matches['query']) ? QueryString::parse($matches['query']) : null,
+            $matches['fragment'] ?? null,
+        );
+    }
+
+    /**
+     * Resolves a relative URI against a base URI.
+     *
+     * Uri::resolve('http://example.com', '/foo/bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo', 'bar'); // http://example.com/foo/bar
```

Comment:
I think this may be 'http://example.com/bar' here

>  return a string consisting of the reference's path component
      appended to all but the last segment of the base URI's path (i.e.,
      excluding any characters after the right-most "/" in the base URI
      path, or excluding the entire base URI path if it does not contain
      any "/" characters).
      
https://www.rfc-editor.org/rfc/rfc3986#page-33